### ISSUE TITLE
Network error when cloning

### ISSUE THREAD

**Issue by sebbacon** [opensafely/job-runner#5]
Sometimes when cloning (e.g. with `git clone --depth 1 --branch master https:/github.com/opensafely/nsaids-research /e/cohorts/tmp1234/nsaids-research-master`) we get

    error: RPC failed; curl 56 GnuTLS recv error (-9): A TLS packed with unexpected length was received

There are a number of issues floating around StackOverflow relating to this, most of them sound like dead-chicken-waving - [this](https://stackoverflow.com/questions/38378914/git-error-rpc-failed-curl-56-gnutls) is a good example with a number of such chickens

Suggestions include replacing the TLS library, changing buffer sizes, etc. I don't think any of them are accurate.

However, this error has so far been transient, so we should retry the checkout a few times in the code.

**Comment by sebbacon** [comment]
So I've been able to pin this down. It is a very particular combination of Docker on Windows combined with a specific git repo state being pulled down.  I say this about the state because I've seen the problem go away following commits being made, but not this time.

* With the repo at exactly [this point](https://github.com/opensafely/carehomes-research/commit/46ba6d026c5fcaf21efb168009b154470141b62a)
* Inside a `job-runner` docker image on Windows on the L3 server
  * But not when in `git bash` on Windows 
  * Aand not when in `job-runner` docker image in Linux
* When running `git clone --depth 1 https://github.com/opensafely/carehomes-research.git`
  * But *not* `git clone https://github.com/opensafely/carehomes-research.git`
* Still happening after I add another commit

**Comment by evansd** [comment]
I tried to test this with the steps the new job server uses which is:
```
mkdir testrepo
cd testrepo
git init --bare
git fetch --depth 1 https://github.com/opensafely/carehomes-research.git 46ba6d026c5fcaf21efb168009b154470141b62a
```
And I couldn't reproduce (this was inside a container running the job-runner image).

That's possibly good news, although I also couldn't reproduce the issue using `git clone` because by the time I tried the repo was back in a good state, so it's possible that I wasn't reproducing accurately.

**Comment by sebbacon** [comment]
Can confirm the `init --bare` and `git fetch` approach also works for me

**Comment by sebbacon** [comment]
Further plot thickenage: our `docker.opensafely.org/r` image does not exhibit the problem.

The one that works is Ubuntu 18.04.3 and the one that's broken is 18.04.5; they are both running the same kernel and have the same versions of various TLS libraries installed.

**Comment by ghickman** [comment]
I've tested this with:

```
#!/bin/sh
mkdir foo
cd foo
git init
git clone --depth 1 https://sebbacon:<PAT>@github.com/opensafely/carehomes-research.git 46ba6d026c5fcaf21efb168009b154470141b62a
```

and I get the gnuTLS error about unexpected packet length.

I also tried it with:

```
#!/bin/sh
mkdir foo
cd foo
git init
git fetch --depth 1 https://sebbacon:<PAT>@github.com/opensafely/carehomes-research.git 46ba6d026c5fcaf21efb168009b154470141b62a
```

which worked.

However, I don't think our original test is doing what we expect.  `git-clone` can't clone a given SHA, only a branch or tag (with `--branch`) so currently we're cloning the full repo into a directory named after the SHA.

**Comment by ghickman** [comment]
💡 sudden evening inspiration: could this be packfile/object related?

And does a local repack of objects transfer to GitHub?

**Comment by sebbacon** [comment]
So I had a half-baked memory of seeing this back in April associated with very large binary files.

The repo did include a shapefile which is 15Mb. I was able to replace this with a much smaller equivalent (3Mb) and I've completely blown away the git history too as a quick way of altering packfiles.

I took a copy as a fork to my personal github account in case we wanted any history and/or something to investigate further.

So far I haven't seen the issue again
